lib/index.js

```diff
@@ -48,49 +50,59 @@ const argv = process.argv[1] || /* c8 ignore next -- windows */ ''
 const nvm = process.env.NVM_BIN
 
 const config = new NpmConfig({
-  definitions: {},
-  npmPath: fileURLToPath(new URL('.', import.meta.url))
+  argv: [],
+  definitions,
+  npmPath: ''
 })
 
-config.loadGlobalPrefix()
-
 // Note: this is a file path.
 // If there is no prefix defined,
 // use the defaults.
 // See: <https://github.com/eush77/npm-prefix/blob/master/index.js>
-/* c8 ignore next 6 -- typically defined */
-/** @type {string} */
-const npmPrefix =
-  config.globalPrefix ||
-  (windows
-    ? path.dirname(process.execPath)
-    : path.resolve(process.execPath, '../..'))
-
-const defaultGlobal = electron || argv.startsWith(npmPrefix)
+/** @type {string | Promise<string>} */
+// eslint-disable-next-line unicorn/prefer-top-level-await
+let npmPrefix = config.load().then(() => {
+  npmPrefix =
+    /* c8 ignore next 6 -- typically defined */
+    config.globalPrefix ||
+    (windows
+      ? path.dirname(process.execPath)
+      : path.resolve(process.execPath, '../..'))
+  return npmPrefix
+})
+
 /** @type {Readonly<LoadOptions>} */
 const defaultLoadOptions = {}
 /** @type {Readonly<ResolveOptions>} */
 const defaultResolveOptions = {}
 
-/* c8 ignore next -- windows */
-const nodeModules = windows ? 'node_modules' : 'lib/node_modules'
-
-let globalFolder = new URL(nodeModules, pathToFileURL(npmPrefix) + '/')
-
-// If we’re in Electron,
-// we’re running in a modified Node that cannot really install global node
-// modules.
-// To find the actual modules,
-// the user has to set `prefix` somewhere in an `.npmrc` (which is picked up
-// by `@npmcli/config`).
-// Most people don’t do that,
-// and some use NVM instead to manage different versions of Node.
-// Luckily NVM leaks some environment variables that we can pick up on to try
-// and detect the actual modules.
-/* c8 ignore next 10 -- Electron. */
-if (electron && nvm && !fs.existsSync(globalFolder)) {
-  globalFolder = new URL(nodeModules, pathToFileURL(nvm))
-}
+/** @type {URL | Promise<URL>} */
+// eslint-disable-next-line unicorn/prefer-top-level-await
+let globalFolder = /** @type {Promise<string>} */ (npmPrefix).then(
+  (npmPrefix) => {
+    /* c8 ignore next -- windows */
+    const nodeModules = windows ? 'node_modules' : 'lib/node_modules'
+
+    globalFolder = new URL(nodeModules, pathToFileURL(npmPrefix) + '/')
+
+    // If we’re in Electron,
+    // we’re running in a modified Node that cannot really install global node
+    // modules.
+    // To find the actual modules,
+    // the user has to set `prefix` somewhere in an `.npmrc` (which is picked up
+    // by `@npmcli/config`).
+    // Most people don’t do that,
+    // and some use NVM instead to manage different versions of Node.
+    // Luckily NVM leaks some environment variables that we can pick up on to try
+    // and detect the actual modules.
+    /* c8 ignore next 10 -- Electron. */
+    if (electron && nvm && !fs.existsSync(globalFolder)) {
```

Comment:
If everything is promises anyway, then we can also turn the whole thing into a nice function, and use the async version of `exists` here, right?

```
let globalFolder = inferGlobalFolder()
```

And then at the bottom:

```
async function inferGlobalFolder() {
  // ...
}
```

Reply:
Well, then we could try and read and see the error that is thrown when that doesn’t work!

> IIRC it’s a bad idea for keep promises around

The `npmPrefix` variable *is* currently still a promise, right?

I think you’re correct. But what are the alternatives? We are memo-izing work because we assume (reasonably) that the global folder doesn’t change. We need to store some value? And we do use a promise?

How about, global scope:


```js
/** @type {string | undefined} */
let globalFolder
```

In `resolvePlugin`:


```js
    if (globals) {
      if (!globalFolder) globalFolder = await inferGlobalFolder
      from.push(globalFolder)
    }
```

And then:

```js
async function inferGlobalFolder() {
 // ...
}
```

Reply:
[`fs.exists`](https://nodejs.org/api/fs.html#fsexistspath-callback) doesn’t throw or log anything, it was just deprecated in Node.js 1.0.0. There’s also no promise version.

`npmPrefix` is handled the same as `globalFolder`. It’s reassigned in the `.then()` callback. No promises are kept around.

Reply:
https://github.com/wooorm/load-plugin/pull/20

I meant to put all this logic into proper named functions. See bottom.
And also, to use the function that Node docs suggest instead of `exists`: `access`.

Indeed, those `from.push`s might be better! I just pushed it in the second commit: then we don’t need `fs` at all!